### PR TITLE
Use uniform ABI for all runtimes

### PR DESCRIPTION
**Pull request by earhart**

### CODE REVIEW

**Comment by earhart** [comment]
Updated to simplify things a bit (there's some scaffolding, but it can be removed in the next PR).  PTAL?

**Comment by earhart** [comment]
After simplifying things, this PR was looking a little light; here's the next phase (which also removes the scaffolding).

File: pmlc/transforms/make_entrypoints.cc [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    signalPassFailure();
+  }
+
+  return entryFunc;
+}
+
+void MakeEntrypointsPass::makeSetup(mlir::OpBuilder &builder) {
+  // This method builds:
+  //
+  //   Device* setup(Device* device) {
```
Comment (flaub) [inline pmlc/transforms/make_entrypoints.cc:86]: What's the return-value for?

File: pmlc/util/ids.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2020 Intel Corporation
+#pragma once
+
+// A collection of well-known identifiers used in codegen.
+
+namespace pmlc {
```
Comment (flaub) [inline pmlc/util/ids.h:6]: `pmlc::util`

File: pmlc/transforms/passes.td [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#ifndef __PMLC_TRANSFORM_PASSES__
```
Comment (flaub) [inline pmlc/transforms/passes.td:1]: Not sure this belongs in a top-level `transforms` directory. Seems like it's `rt`-specific

File: pmlc/rt/jit_executable.cc [hunk 4]
```diff
@@ -428,7 +458,8 @@ class JitExecutable final : public Executable {
     descriptors.clear();
     ptrs.clear();
 
-    std::copy(preParams.begin(), preParams.end(), std::back_inserter(ptrs));
+    network = abi.setupFunc(this->device.get());
```
Comment (flaub) [inline pmlc/rt/jit_executable.cc:461]: We need to pass all the user arguments to the setup function also.

Comment (flaub) [reply pmlc/rt/jit_executable.cc:461]: Actually, just the constant buffers need to be passed

File: pmlc/transforms/make_entrypoints.cc (hunk 1 above)
Comment (earhart) [reply pmlc/transforms/make_entrypoints.cc]: Added a comment.

File: pmlc/util/ids.h (hunk 2 above)
Comment (earhart) [reply pmlc/util/ids.h]: Done.  (Seems unnecessary to me, tho -- lots of `util::` sprinkled everywhere for no obvious value.)

File: pmlc/transforms/passes.td (hunk 3 above)
Comment (earhart) [reply pmlc/transforms/passes.td:1]: Subdirectories of `rt` are runtimes, and the pass is used by the target pipelines, not`rt` itself -- it's a way to implement the expected `rt` ABI, and not even the only way.

File: pmlc/rt/jit_executable.cc (hunk 4 above)
Comment (earhart) [reply pmlc/rt/jit_executable.cc:463]: Definitely.  But I'd rather do that in a followup PR.

File: pmlc/util/ids.h [hunk 5]
```diff
@@ -3,11 +3,11 @@
...
 // A collection of well-known identifiers used in codegen.
 
-namespace pmlc {
+namespace pmlc::util {
 
 static constexpr const char kEntrypoint[] = "main";
 static constexpr const char kPlaidmlInit[] = "plaidml_init";
-static constexpr const char kPlaidmlExecute[] = "plaidml_exec";
+static constexpr const char kPlaidmlExec[] = "plaidml_exec";
 static constexpr const char kPlaidmlFini[] = "plaidml_fini";
```
Comment (flaub) [inline pmlc/util/ids.h]: I feel like these don't belong in `util`? Maybe these are `rt` specific?

File: pmlc/util/ids.h (hunk 2 above)
Comment (flaub) [reply pmlc/util/ids.h]: It's just consistency with the directory structure. I'm fine with flattening the namespace so it's not so deep, but I think we should do that with buy-in from others and that should be its own NFC PR.

File: pmlc/ast/builder.cc [hunk 6]
```diff
@@ -38,8 +39,6 @@ namespace tile = pmlc::dialect::tile;
 
 namespace {
 
-static constexpr const char *kEntrypoint = "main";
```
Comment (flaub) [inline pmlc/ast/builder.cc:41]: Not sure we should move this, seems like it's only local to the builder.

File: pmlc/rt/jit_executable.cc (hunk 4 above)
Comment (flaub) [reply pmlc/rt/jit_executable.cc:463]: I think we should call `setupFunc` in the constructor, not in `bindArguments`, but I suppose that can wait in a follow up.

File: pmlc/transforms/make_entrypoints.cc [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  auto ctx = &getContext();
+  llvmInt32Type = LLVMType::getInt32Ty(ctx);
+  llvmPtrTy = LLVMType::getInt8PtrTy(ctx);
+  llvmVoidTy = LLVMType::getVoidTy(ctx);
+}
+
+mlir::FuncOp MakeEntrypointsPass::getNetworkEntry() {
+  mlir::FuncOp entryFunc;
+
+  getOperation().walk([&](mlir::FuncOp func) {
```
Comment (flaub) [inline pmlc/transforms/make_entrypoints.cc:58]: Don't use walk, use `getOps<FuncOp>` so that we don't look for recursive functions.

File: pmlc/transforms/make_entrypoints.cc [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  mlir::FuncOp entryFunc = getNetworkEntry();
+  if (!entryFunc) {
+    return;
+  }
+
+  initMembers();
+  mlir::OpBuilder builder(getOperation().getBody()->getTerminator());
+  makePlaidmlInit(builder);
+  makePlaidmlExecute(builder, entryFunc);
+  makePlaidmlFini(builder);
```
Comment (flaub) [inline pmlc/transforms/make_entrypoints.cc:45]: We should check to see if these functions are already defined or at least fail if they're already defined.

File: pmlc/transforms/make_entrypoints.cc [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  //
+  //   Network* plaidml_init(Device* device) {
+  //     return device;  // Network == Device for now.
+  //   }
+  //
+  // This is a rather trivial function; the plan is to expand it later.
+  // In particular, the Network parameter gives networks a mechanism to carry
+  // additional information from plaidml_init to plaidml_exec and
+  // plaidml_fini.
+  auto func = builder.create<mlir::FuncOp>(
```
Comment (flaub) [inline pmlc/transforms/make_entrypoints.cc:91]: I'd expect to be operating in the LLVM dialect and to create `llvm::FuncOp` instead of `mlir::FuncOp`

File: pmlc/transforms/make_entrypoints.cc [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    entryFunc = func;
+    return mlir::WalkResult::advance();
+  });
+  if (!entryFunc) {
+    getOperation().emitError(
+        "Expected a single entrypoint function definition");
+    signalPassFailure();
+  }
+
+  return entryFunc;
```
Comment (flaub) [inline pmlc/transforms/make_entrypoints.cc:77]: Why don't we just use a symbol lookup?

File: pmlc/transforms/passes.td [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+#define __PMLC_TRANSFORM_PASSES__
+
+include "mlir/Pass/PassBase.td"
+
+def MakeEntrypointsPass : Pass<"pmlc-make-entrypoints", "mlir::ModuleOp"> {
```
Comment (flaub) [inline pmlc/transforms/passes.td:6]: It seems like this pass should operate at the llvm dialect level, so I'd call it `llvm-make-entrypoints` or perhaps some better name (can't think of one right now).

File: pmlc/transforms/make_entrypoints.cc (hunk 9 above)
Comment (flaub) [reply pmlc/transforms/make_entrypoints.cc:91]: Alternatively, I could see this split happening at PXA dialect or comp dialect level, and then having a straightforward legalization to convert high-level `mlir::FuncOp` into low level `llvm::FuncOp` (converting a `Network` data type in to a `void*` for instance).

File: pmlc/rt/jit_executable.cc [hunk 12]
```diff
@@ -388,26 +418,28 @@ class JitExecutable final : public Executable {
...
     }
 
-    jitEntry = impl->compile(std::move(llvmModule), std::move(ctx), entryPoint);
-    if (!jitEntry) {
-      throw std::runtime_error("jitEntry function is null");
+    abi = impl->compile(std::move(llvmModule), std::move(ctx));
+    if (!abi) {
+      throw std::runtime_error("Entrypoint functions not found");
     }
   }
```
Comment (flaub) [inline pmlc/rt/jit_executable.cc:431]: Do reduce duplication, let's build up the ABI here, after `impl->compile` so it's part of the common code path.

File: plaidml/edsl/tests/BUILD [hunk 13]
```diff
@@ -29,7 +29,7 @@ plaidml_cc_test(
         "//plaidml:intel_gen_ocl_spirv": [":edsl_test-skip.intel_gen_ocl_spirv.txt"],
         "//conditions:default": [],
     }),
-    linkstatic = True,
+    # linkstatic = False,
```
Comment (flaub) [inline plaidml/edsl/tests/BUILD:32]: Probably need to undo this

File: pmlc/conversion/abi_to_llvm/abi_to_llvm.cc [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+#include "pmlc/conversion/abi_to_llvm/passes.h"
+#include "pmlc/dialect/abi/ir/dialect.h"
+#include "pmlc/util/ids.h"
+
+// #pragma GCC diagnostic ignored "-Wunused-function"
+// #pragma GCC diagnostic ignored "-Wunused-variable"
```
Comment (flaub) [inline pmlc/conversion/abi_to_llvm/abi_to_llvm.cc:17]: I don't think this is supported on all compilers, why do we need these?

File: pmlc/conversion/abi_to_llvm/abi_to_llvm.cc [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+    builder.setInsertionPointToStart(moduleOp.getBody());
+    func =
+        builder.create<LLVM::LLVMFuncOp>(builder.getUnknownLoc(), name, funcTy);
+  }
+  return func;
+}
+
+namespace {
+
+constexpr char emptyNetworkSingletonName[] = "plaidml_empty_network";
```
Comment (flaub) [inline pmlc/conversion/abi_to_llvm/abi_to_llvm.cc:40]: Rename to `kNetworkSingletonName`

File: pmlc/conversion/abi_to_llvm/input.mlir [hunk 16]
```diff
@@ -0,0 +1,55 @@
+Converting fn: llvm.func @main(%arg0: !llvm.ptr<float>, %arg1: !llvm.ptr<float>, %arg2: !llvm.i64, %arg3: !llvm.i64, %arg4: !llvm.i64, %arg5: !llvm.i64, %arg6: !llvm.i64, %arg7: !llvm.ptr<float>, %arg8: !llvm.ptr<float>, %arg9: !llvm.i64, %arg10: !llvm.i64, %arg11: !llvm.i64, %arg12: !llvm.i64, %arg13: !llvm.i64, %arg14: !llvm.ptr<float>, %arg15: !llvm.ptr<float>, %arg16: !llvm.i64, %arg17: !llvm.i64, %arg18: !llvm.i64, %arg19: !llvm.i64, %arg20: !llvm.i64) {
```
Comment (flaub) [inline pmlc/conversion/abi_to_llvm/input.mlir:1]: Can we make this a lit test?

File: pmlc/dialect/abi/ir/dialect.h [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+using mlir::StringAttr;
+using mlir::StringRef;
+using mlir::Type;
+using mlir::TypeAttr;
+using mlir::ValueRange;
+using mlir::VectorType;
+
+namespace MemoryEffects = mlir::MemoryEffects;
+namespace SideEffects = mlir::SideEffects;
+*/
```
Comment (flaub) [inline pmlc/dialect/abi/ir/dialect.h:68]: Drop comments

File: pmlc/dialect/abi/ir/loop_op.cc [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+      auto definingOp = val.getDefiningOp();
+      if (definingOp && definingOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+        // N.B. Constant-Like operations have no side-effects, exactly one
+        //      result, and no operands.
+        auto *denormOp = definingOp->clone();
+        rewriter.insert(denormOp);
+        auto arg = loopOp.bodyEntryBlock()->getArgument(idx);
+        arg.replaceAllUsesWith(denormOp->getResult(0));
+        loopOp.bodyEntryBlock()->eraseArgument(idx);
+        networkOp.getOperation()->eraseOperand(idx);
```
Comment (flaub) [inline pmlc/dialect/abi/ir/loop_op.cc:37]: Use `m_Constant` rather than looking for `ConstantLike`

File: tools/conda_run/conda_unix.cc [hunk 19]
```diff
@@ -53,9 +53,9 @@ int main(int argc, char* argv[]) {
 
 #ifdef DEBUG
     // Useful debugging code
-    std::cout << "argc: " << argc << std::endl;
+    std::cerr << "argc: " << argc << std::endl;
```
Comment (flaub) [inline tools/conda_run/conda_unix.cc:56]: Probably doesn't belong in this PR

File: pmlc/rt/jit_executable.cc [hunk 20]
```diff
@@ -411,82 +360,105 @@ class JitExecutable final : public Executable {
...
+      throw std::runtime_error(
+          "Program outputs arguments and buffers mismatch");
+    }
     StopWatch stopWatch;
     if (VLOG_IS_ON(1)) {
       stopWatch.start();
     }
-    bindArguments(inputBuffers, outputBuffers);
-    abi.executeFunc(ptrs.data());
+    auto memrefIt = memrefs.begin();
```
Comment (flaub) [inline pmlc/rt/jit_executable.cc:413]: Rename to `itMemref`

File: pmlc/rt/jit_executable.cc [hunk 21]
```diff
@@ -73,89 +73,41 @@ static void setupTargetTriple(llvm::Module *llvmModule) {
...
-}
-
 namespace {
 
-class MemRefDescriptor {
-private:
-  struct Base {
+class MemRef {
+public:
+  struct Descriptor {
```
Comment (flaub) [inline pmlc/rt/jit_executable.cc:80]: What do we gain by refactoring as such? This seems confusing to me, we want to manipulate a `MemRefDescriptor` not a `MemRef` itself (a memref is just a logical concept).

**Comment by earhart** [comment]
FWIW -- this is currently just a draft PR.  Please feel free to ignore it; if you do want to review, please focus on major architectural issues, since there are many, many small things to be cleaned up before it's ready to submit.

**Comment by flaub** [comment]
SGTM, however it's hard for me to grok what the pass is doing without a lit test. That's probably the first thing I would try to understand when doing an architectural review.

**Comment by earhart** [comment]
+1 -- as things get a little closer to being nailed down, I'll definitely be adding some lit tests.  :-)

**Comment by earhart** [comment]
(I'm actually not all that concerned about the overall architecture -- it's basically the same as what we discussed in the doc review.)

File: pmlc/dialect/abi/ir/loop_op.cc (hunk 18 above)
Comment (earhart) [reply pmlc/dialect/abi/ir/loop_op.cc:37]: Just wondering -- what's wrong with `ConstantLike`?

Comment (flaub) [reply pmlc/dialect/abi/ir/loop_op.cc:37]: Nothing, it's just that `m_Constant` is probably a much more concise way of matching a constant op. But if it's unsuitable, obviously this way is fine :)

**Comment by flaub** [comment]
Is it possible to split up this PR so that the hoisting is done in a follow up? We're going to need to switch to a new task soonish.

**Comment by earhart** [comment]
> Is it possible to split up this PR so that the hoisting is done in a follow up? We're going to need to switch to a new task soonish.

Definitely possible.

File: pmlc/dialect/abi/ir/loop_op.cc [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+
+void LoopOp::getCanonicalizationPatterns(
+    mlir::OwningRewritePatternList &patterns, mlir::MLIRContext *ctx) {
+  patterns.insert<DenormalizeConstantsPattern>(ctx);
+  patterns.insert<RemoveUnusedNetworkFieldsPattern>(ctx);
+}
+
+mlir::Region &LoopOp::getLoopBody() { return bodyRegion(); }
+
+bool LoopOp::isDefinedOutsideOfLoop(mlir::Value value) {
```
Comment (flaub) [inline pmlc/dialect/abi/ir/loop_op.cc:103]: You can probably use `mlir/Transforms/RegionUtils.h`, specifically the `getUsedValuesDefinedAbove` or `getUsedValuesDefinedAbove` helper functions.

Comment (earhart) [reply pmlc/dialect/abi/ir/loop_op.cc:103]: Awesome -- thanks!

**Comment by flaub** [comment]
Instead of writing our own hoisting pass, I think we can use the existing LICM pass that is defined upstream. I realize that your pass also is splitting allocs and deallocs. But I think this could be done in a different pass. I believe we already have a pass that injects deallocs for each alloc. So ideally, this hoisting would occur before we get to the point of adding the deallocs.

**Comment by earhart** [comment]
> Instead of writing our own hoisting pass, I think we can use the existing LICM pass that is defined upstream. I realize that your pass also is splitting allocs and deallocs. But I think this could be done in a different pass. I believe we already have a pass that injects deallocs for each alloc. So ideally, this hoisting would occur before we get to the point of adding the deallocs.

@flaub Initially, this code did use the existing LICM pass.  That pass wound up having two limitations: it wasn't able to move operations to "after the loop" (LoopLike just doesn't define a mechanism for it), and it doesn't move operations with memory effects.

Any post-processing to inject de-allocs would have to know to inject them into the correct location of the fini region of the loop; it's not as trivial as one might like.

Another solution might be to require the execution environment to implicitly destroy kernels and allocs.  That was pretty roundly shot down in the design review, but maybe now that the complexity's a little more obvious, we could go back to that approach?  OTOH, we'd still need to do our own thing for operations with memory effects (or just pretend that the relevant operations don't have effects), and it'd be strictly less general than the existing code.

My own thinking is that we should probably aim to contribute this logic upstream; code review there may suggest additional simplifications.

**Comment by flaub** [comment]
I basically think we should make this a later problem to solve, because we might come up with different solutions over time with varying degrees of simplicity. To me, hoisting is an optimization, not a requirement for legalization/lowering. So I think we should just drop all the hoisting stuff for now and leave that as a follow up PR (or series of PRs).

**Comment by earhart** [comment]
We should definitely split this PR up.

I agree that hoisting is an optimization, not a requirement for legalization and lowering; OTOH, performance is its own requirement.

**Comment by flaub** [comment]
@jbruestle Is this still valid?
